Sources/HTTP/HTTPStreamingParser.swift

```diff
@@ -284,21 +304,29 @@ public class StreamingParser: HTTPResponseWriter {
     func bodyReceived(data: UnsafePointer<Int8>?, length: Int) -> Int32 {
         processCurrentCallback(.bodyReceived)
         guard let data = data else { return 0 }
+        if shouldStopProcessingBody {
+            return 0
+        }
         data.withMemoryRebound(to: UInt8.self, capacity: length) { (ptr) -> Void in
-            let buff = UnsafeBufferPointer<UInt8>(start: ptr, count: length)
+            #if swift(>=4.0)
+                let buff = UnsafeRawBufferPointer(start: ptr, count: length)
+            #else
+                let buff = UnsafeBufferPointer<UInt8>(start: ptr, count: length)
+            #endif
             let chunk = DispatchData(bytes: buff)
             if let chunkHandler = self.httpBodyProcessingCallback {
-                var stop = false
-                var finished = false
-                while !stop && !finished {
-                    switch chunkHandler {
+                switch chunkHandler {
                     case .processBody(let handler):
-                        handler(.chunk(data: chunk, finishedProcessing: {
-                            finished = true
-                        }), &stop)
+                        //OK, this sucks.  We can't access the value of the `inout` inside this block
+                        //  due to exclusivity. Which means that if we were to pass a local variable, we'd
+                        //  have to put a semaphore or something up here to wait for the block to be done before
+                        //  we could get its value and pass that on to the instance variable. So instead, we're
+                        //  just passing in a pointer to the internal ivar. But that ivar can't be modified in
+                        //  more than one place, so we have to put a semaphore around it to prevent that.
+                        _shouldStopProcessingBodyLock.wait()
+                        handler(.chunk(data: chunk, finishedProcessing: {self._shouldStopProcessingBodyLock.signal()}), &_shouldStopProcessingBody)
```

Comment:
`&_shouldStopProcessingBody` isn't protected by the lock....

Reply:
Do we actually need to wait for the async completion handler to run before we can set stop? Don't we want to set it immediately (sync) even if an async `writeBody()` is still going to happen?

Reply:
> &_shouldStopProcessingBody isn't protected by the lock....

?
That's what line 326 does.

> Do we actually need to wait for the async completion handler to run before we can set stop? Don't we want to set it immediately (sync) even if an async writeBody() is still going to happen

We can't set `stop` here - it's passed as in `inout` to the API, and it's set by the `handler`.

Reply:
The completion handler that releases the lock runs async (separate thread) - there's no guarantee that its going to call signal() after `_shouldStopProcessingBody` is modified, so its effectively unprotected.

I was wondering whether your intent was to block access to `_shouldStopProcessingBody` until after the completion handler has run.